### DISCUSSION TITLE
[A-L] (10 settembre 2024) Esercizio 1

### DISCUSSION THREAD

**Discussion by FeddyLix17** [sapienzastudentsnetwork/progettazione-di-algoritmi#114] [Esami A-L]
Per un grafo diretto G, diciamo che un arco (u, v) è *interno* se u e v appartengono alla stessa componente fortemente connessa e altrimenti diciamo che è *esterno*.

In relazione ad un DFS qualsiasi, ripsondere alle seguenti domande:

- (1) un arco all'indietro può essere esterno?

- (2) un arco in avanti può essere interno?

- (3) un arco di attraversamento può essere interno?

- (4) un arco di attraversamento deve essere esterno?

Per ognuna, in caso affermativo esibire un esempio e altrimenti dimostrare l'impossibilità.

**Comment by Viquess** [comment]
Stavo ragionando sul punto 3 e ho realizzato un grafo che, credo, fornisca un esempio di arco d'attraversamento interno. Tuttavia il punto 4 immediatamente successivo mi fa pensare di aver sbagliato qualcosa nel ragionamento.

Immaginando un grafo del genere:
![image](https://github.com/user-attachments/assets/8e74e79c-bab6-4f6e-a14d-0208252c5674)

Le SCC dovrebbero essere:
1. {1}
2. {2, 3, 4, 5}

Ipotizziamo ora una DFS che visita i nodi in questo ordine: 1, 2, 3, 4, 5... Ovvero visita i nodi da 1 a 4, poi torna indietro fino al 2 e visita il nodo 5.

A questo punto l'arco (5, 4) dovrebbe essere un arco di attraversamento, dato che quando 5 viene visitato, 4 è già stato visitato completamente.

Ci troveremmo quindi davanti ad un arco di attraversamento interno, in quanto i nodi 4 e 5 fan parte della stessa SCC.
Sbaglio qualcosa nel ragionamento?

**Reply by Near047** [reply]
Non saprei perchè in quel caso l'SCC dovrebbe essere {2,3,4}

**Reply by Viquess** [reply]
> Non saprei perchè in quel caso l'SCC dovrebbe essere {2,3,4}

Come mai?
Teoricamente una SCC si definisce come un sottografo in cui per ogni coppia di nodi u, v esiste un cammino che li connette.
E in questo caso esistono cammini che connettono 5 a 2, 3 e 4 e viceversa.

**Reply by AshBaw1** [reply]
Penso che il tuo esempio sia corretto. Un altro esempio potrebbe essere il grafo con nodi {A, B, C} e archi {(A, B), (B, A), (A, C), (C, B)}.

SCC: {A, B, C} (tutti mutuamente raggiungibili).

DFS con ordine A → B → C:

C → B è un arco di attraversamento.

**Comment by cek753** [comment]
1- no, non può esssere esterno. Per assurdo se (u.v) esterno all'indietro, questo vuol dire che u e v non fanno parte della stessa comp connessa, ma se è un arco all'indietro vuol dire che ho un modo per andare da v a u CONTRADDIZIONE!(fanno parte della stessa comp connessa).

2- si, basti pensare ad un grafo con vertici {1,2,3} e archi {(1,2),(2,3),(1,3),(3,1)} , supponiamo che la dfs visiti gli archi in questo ordine  (1,2)(2,3)(1,3) abbiamo che (1,3) è un arco in avanti e il grafo è fortemente connesso

3-si, per empio se si hanno due alberi che partono dalla stessa radice, un arco di attraversamento tra questi due alberi andrebbe a creare un'unica comp connessa

4-no, supponiamo per assurdo che l'arco (u,v) sia esterno, in questo modo u e v fanno parte di due diverse componenti connese,però, supponendo che esiste un (x,y) che connette le due componenti connesse  con x appartentente alla comp connessa di v e y a quella di u,  in questo modo si andrebbe a creare una comp connessa unica -><- CONTRADDIZIONE
